docs/customization.rst

```diff
@@ -429,6 +429,63 @@ Configuration Options
         specified by :confval:`sphinx_immaterial_external_resource_cache_dir`)
         and included in the built documentation.
 
+        To use the user's system font instead of (and similar to) the theme's default fonts,
+        set this field to :python:`False`.
+
+        .. code-block:: python
+            :caption: rely on fonts available in the user's OS.
+
+            html_theme_options = {
+                "font": False,
+            }
+
+        Using a custom or self-hosted font requires extra CSS to define the font family and
+        overridden CSS theme variables. The following steps were used to self-host Google's
+        "Comic Neue" and "Comic Mono" fonts as an example.
+
+        1. Copy the font files to your project's ``_static`` folder. The font's file
+           :css:`format()` will be specified in step 2. If the font has different weights, then
+           each weight will have a separate file. By default, this theme mainly uses 300, 400, and
+           700 weights.
+        2. Define the font family in a CSS file and append the CSS file's name to the project's
+           list of `html_css_files` in config.py.
+
+           Each weight must have a corresponding :css:`@font-face` rule. The same mandate applies
+           to *italic* variants (if any) and their weights.
+
+           .. literalinclude:: _static/custom_font_example.css
+               :caption: docs/_static/custom_font_example.css
+               :language: css
+               :start-at: /* comic-neue-latin-400-normal
+               :end-before: /* comic-neue-latin-700-normal
+           
+           The path specified in the :css:`url()` is relative to the file defining the font family.
+           For example, the above snippet is located in this project's ``_static`` folder where the
+           font's files are located in ``_static/example-custom-font/comic_*/`` folders.
+        3. Override the theme's CSS variables.
+
+           .. code-block:: css
+               
+               :root {
+                 --md-text-font: "Comic Neue"; /* (1)! \*/
+                 --md-code-font: "Comic Mono"; /* (2)! \*/
+               }
+           
+           .. code-annotations::
+                1. Used for regular text.
+                2. Used for ``code snippets``.
+
+           .. warning::
+               Always define fonts with the CSS variables shown above and not with
+               :css:`font-family` because :css:`font-family` uses the theme's CSS variables that
+               define the system font fallback.
+
+        .. rst-example::
+            :class: custom-font-example
+
+            This text is just an **example**.
+            *Notice* the font family used is different (monospaced) in the code snippet.
+
```

Comment:
Perhaps add a note that the graphviz extension won't use the custom font.

Perhaps we should also allow the user to continue using `html_theme_options['font']` to set the fonts even when using custom fonts, and we should just have a config option to disable injecting the google font?  Then the user wouldn't need the extra `:root {` font block that kind of exposes implementation details of the theme.

Reply:
> Perhaps add a note that the graphviz extension won't use the custom font.

Will do.

> Perhaps we should also allow the user to continue using `html_theme_options['font']` to set the fonts even when using custom fonts

That's what the docs are actually doing for the limited example.

> we should just have a config option to disable injecting the google font?

Isn't that the same as using `html_theme_options['font'] = False`? I'm kinda hazy on how the `False` value is handled, but it does seem to work as expected in upstream.

> Then the user wouldn't need the extra `:root {` font block that kind of exposes implementation details of the theme.

I think the more important implementation detail is overriding the `--md-*-font-family` CSS vars as that is where fallback fonts are specified. I added the warning about not overriding the `font-family` property because it basically uses the `--md-*-font-family` value.

In creating the docs' example. I noticed the CSS specificity was an issue for a narrow scope (of custom/self-hosted font). I left a comment in the extra_css.css mods to explain why I had to override the `--md-*-font-family` vars and the `font-family` property... It works fine if the `--md-*-font` vars are altered at the doc's `:root` scope, but the example I used in the docs seems like a edge case exception.

---
Since this approach is the one recommended upstream, I think it'll do for now. But I would like to consider an idea that removes all the boilerplate CSS with some extra fields like
```py
html_theme_options = {
    "font": {
        "custom_text": "font-name",
        "custom_code": "mono-font-name",
    }
}
```
or maybe something similar to our custom admonitions config.

Reply:
Well here it is unusual since you are just using the custom font in one place.  But I'm imagining that we have something like:

`sphinx_immaterial_inject_google_font = False`

that would prevent us from injecting the font specified by `html_theme_options["font"]`.  However, the font family would still get set to the specified fonts in the normal way, so they should either be system fonts or defined by the user with custom css as in your example.

It is true this wouldn't save much but it would mean the user wouldn't need to create a custom css file just to use a system font.

As for `"custom_text"` and `"custom_code"` --- did you have in mind exactly what I'm describing as far as `sphinx_immaterial_inject_google_font = False`, or do you mean somehow also adding css rules to define the font?

Reply:
> the user wouldn't need to create a custom css file just to use a system font.

Oops, I should've made this clearer. No custom CSS is needed for using system fonts; only `"font": False` is needed for that. Or are you referring to a project that may want to use a system font not specified in the theme's fallback options? Because that would simply require the system fonts' name in `:root { --md-*-font: "name"; }`.

> or do you mean somehow also adding css rules to define the font?

Yes. One of the hardships I encountered when putting the example together was the CSS `url()`  path to the font file. I tried to make as much explicit in the docs, but there is clearly possible points of error in upstream's recommended approach.

Reply:
Indeed it expects a custom.css when I set `font: false`. But not a blocker during build it will only try to load custom.css when accessing it and gives a 404 in developer mode.
I solved that by placing an empty `_static/custom.css`.

Reply:
Again with the useful feedback! I didn't notice a 404 in the dev console. I will have to investigate and amend the docs about this.

@hille721 were you using chrome? I mainly just test in firefox...

Reply:
Edge, thus kind of Chrome.

Reply:
That is odd -- not sure where `custom.css` is coming from....

Reply:
In loading the generated docs from this PR's added unit test, I don't see any 404 error. I do see a JS error (in all browsers: edge chrome, firefox), but that JS error is related to the unit test having a bare minimal config (namely an undefined `html_theme_options["palette"]`).

@hille721 You might want to check your conf.py for a `html_css_files` setting. If that is set to `["custom.css"]`, then I would expect the docs' src folder to have a  `_static/custom.css` file.

Reply:
@2bndy5 : ok my bad, not directly related to sphinx-immaterial. It seems to be somehow a problem when combining sphinx-immaterial with [sphinx-argparse-cli](https://pypi.org/project/sphinx-argparse-cli/). Was able to reproduce in a fresh environment:

``` python
project = 'test'
copyright = '2023, ch'
author = 'ch'
extensions = [
    "sphinx_argparse_cli",
    "sphinx_immaterial"
    ]
templates_path = ['_templates']
exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
html_static_path = ['_static']
html_theme = 'sphinx_immaterial'
```

``` bash
sphinx==6.1.3
sphinx-argparse-cli==1.11.0
sphinx-immaterial==0.11.3
```

![image](https://user-images.githubusercontent.com/38468371/221814536-a2f005fb-d51e-4b7c-bb6d-dd5ff54e10c4.png)

Reply:
Yep, I [found the problem in that ext's src](https://github.com/tox-dev/sphinx-argparse-cli/blob/4a542ff54cb34e6400fece0e320e37025d4514bf/src/sphinx_argparse_cli/__init__.py#L9). Although I don't see a place where the `custom.css` is generated or distributed in that ext. It is worth [opening an issue there](https://github.com/tox-dev/sphinx-argparse-cli/issues), but the repo doesn't look very active.

Reply:
I think `custom.css` is normally coming with plain Sphinx (alabster):

![image](https://user-images.githubusercontent.com/38468371/221821631-1d2a54af-3f3f-470c-a047-5005dc1e272b.png)

Seems to be they are providing kind of a placeholder `custom.css`.

Will open a issue there as I also just realized that this problem occurs also when using the popular rtd theme.

Reply:
hold up!

We actually don't inherit CSS from sphinx basic theme's (like many other popular sphinx theme's actually do), so we may need to consider a solution for this on our end.

Reply:
I still don't see where custom.css is coming from the sphinx codebase. I do see a few assertions that the file exists in some of sphinx' unit tests, but I'm having trouble tracking it down.

Either way, I'm not convinced this is a sphinx-immaterial problem.

Reply:
yeah as I said, problem occurs also with rtd theme. Already created a issue there (https://github.com/tox-dev/sphinx-argparse-cli/issues/82)

Reply:
I think at some point sphinx abstracted the alabaster theme as a dependency...

Reply:
Ok, so getting back on track here...

I added the note about the graphviz ext in the the graphviz.rst doc (with a link back to the `:themeconf:'font'` doc) when not using a Google font.

As for the rest, I think we were beginning to discuss a more user-friendly feature about specifying a custom or self-hosted font. But I feel like there was a disconnect between written word and my own comprehension. Maybe we should start a new issue to discuss in detail?